pkg/kritis/review/review.go

```diff
@@ -61,8 +61,35 @@ func generateGapAllowlist(gaps []v1beta1.GenericAttestationPolicy) []string {
 	return allowlist
 }
 
+// ReviewImageWithGAP reviews single image against a single generic attestation policy
+func (r Reviewer) ReviewImageWithGAP(image string, gap v1beta1.GenericAttestationPolicy, c metadata.ReadOnlyClient) (bool, []string, error) {
+	glog.Infof("Validating against GenericAttestationPolicy %s", gap.Name)
+
+	// Get all AttestationAuthorities in this policy.
+	auths, err := r.getAttestationAuthoritiesForGAP(gap)
+	if err != nil {
+		return false, nil, err
+	}
+	if len(auths) == 0 {
+		return false, nil, fmt.Errorf("Generic attestation policy %s did not specify any attestation authority.", gap.Name)
+	}
+
+	notAttestedBy, _ := r.findUnsatisfiedAuths(image, auths, c)
+	if len(notAttestedBy) == 0 {
+		glog.Infof("Image is attested by gap %s", gap.Name)
+		return true, make([]string, 0), nil
+	}
+
+	notAttestedAuthNames := make([]string, len(notAttestedBy))
+	for i, notAttestedAuth := range notAttestedBy {
+		notAttestedAuthNames[i] = notAttestedAuth.Name
+	}
+	glog.Infof("Image is missing attestations from %+q in gap %s", notAttestedAuthNames, gap.Name)
+
+	return false, notAttestedAuthNames, nil
+}
+
 // ReviewGAP reviews images against generic attestation policies
-// Returns error if violations are found and handles them per violation strategy
 func (r Reviewer) ReviewGAP(images []string, gaps []v1beta1.GenericAttestationPolicy, pod *v1.Pod, c metadata.ReadOnlyClient) error {
 	images = util.RemoveGloballyAllowedImages(images)
```

Comment:
I would do the "are there any aas" check first, and error out if not.  IMO, this will result in less surprising behavior.  (Because right now, whether a policy error is returned depends on the images being reviewed, not only on the policy itself.)